### ISSUE TITLE
P2 Sampling

### ISSUE THREAD

**Issue by nkalyanv99** [pengzhangzhi/Open-dLLM#7]
Hi,

I just wanted to clarify a possible error in the P2 sampling code, from what I've understood in the code quoted below, the confidence scores are only calculated for the masked tokens, and so all tokens that are already unmasked have infinite scores and will never be remasked. 

`             confidence, x0 = sample_tokens(mask_logits, temperature, top_p, top_k, alg=alg)
                # Scatter confidence back to full sequence
                full_conf = torch.full_like(x, float("inf"), dtype=confidence.dtype)
                full_conf[mask_index] = confidence
`

I think this deviates from the original paper, where we want to also remask unmasked tokens too and (confidence, x0) should be calculated for everything but the prompt?

**Comment by pengzhangzhi** [comment]
hey thank you soo much!!!

https://github.com/pengzhangzhi/Open-dLLM/blob/8aba92eb51cbe0c8ff2e70bf44416dac8ae7507a/veomni/models/transformers/qwen2/generation_utils.py#L235

You are right, this is a mistake, we should get rid of this line, and set the `full_conf=confidence` and only set the `full_conf[fix_mask] = float("inf")` 

so instead of doing:
```
# wrong 
full_conf = torch.full_like(x, float("inf"), dtype=confidence.dtype)
full_conf[mask_index] = confidence
full_conf[fix_mask] = float("inf")


# correct
full_conf = confidence
full_conf[fix_mask] = float("inf")

```

What do you think?

**Comment by nkalyanv99** [comment]
Hey thanks for the quick reply :) 

I think the logits should be computed over all tokens but the fixed mask? 
So it can be simplified to:

`confidence, x0 = sample_tokens(logits, temperature, top_p, top_k, alg=alg) #Note logits not masked_logits
confidence[fix_mask] = float('inf')`. 

And then the rest works from there?

**Comment by pengzhangzhi** [comment]
ya we might be saying the same thing lol.

Ty for spotting this!! Do u want to make a PR? This is an important fix lol

**Comment by nkalyanv99** [comment]
Sure, would be happy to :)

**Comment by nkalyanv99** [comment]
Made and partially benchmarked the changes, but there's some style and make errors that I think Alex has fixed in his PR. So for now I'll wait for that to be merged first?

**Comment by pengzhangzhi** [comment]
lol probably not? Alex might not have time for that and leaves the draft there forever lol.

**Comment by nkalyanv99** [comment]
Oops lol I somehow missed it was a draft. I saw so many changes I thought it was done, let me send it in.

**Comment by nkalyanv99** [comment]
Ah I didn't see Zhang already did it, closed the PR and closing the issue, since it's fixed now :)

**Comment by pengzhangzhi** [comment]
yep he did tyty for spotting this!!
